doc/designs/evp_skey.md

```diff
@@ -0,0 +1,175 @@
+EVP_SKEY Design
+===============
+
+We currently presume that symmetric keys are represnted as octet buffer. It may
+be not relevant for the situation of hardware-backed providers, such as
+PKCS#11 or TPM.
+
+This is a design of using opaque usage in symmetric encryption.
+
+Provider perspective
+--------------------
+
+On the provider end, encryption and decryption implementations
+would receive an opaque pointer similar to the provkey pointer used with
+asymmetric operations. They expected to know how to deal with the structure.
+
+The structure is the result of processing an array of OSSL_PARAM on the
+provider level. The exact set of those params is provider-specific.
+
+It should be possible to export/import the raw algorithm-specific
+keydata of an EVP_SKEY from a provider to another provider. Of course a
+provider could refuse the export operation if the key is unexportable.
+
+The EVP_SKEY object *can* have an algorithm association. It should be a
+provider-specific decision or maybe an information that could be associated to
+an EVP_SKEY any time later. I.e., it starts as algorithm unspecific and may
+later transform into an algorithm specific object by some additional setter.
+
+We do not see why we would want to assign an algorithm immediately during
+the creation of an EVP_SKEY as result of derive() operation, if the operation
+just creates bunch of bytes that can be used as a key equally well for HMAC or
+AES encryption.
+
+So, listening to what's been said, it sounds like an EVP_SKEY would
+conceptually be a union of "a bunch of bytes" (with no particular algorithm in
+mind) and a reference to a provider-backed symmetric key (which will have to be
+specific to a key type / base cipher algo).
```

Comment:
Maybe such a direct quote isn't the best.  How about this instead:

``` markdown
To allow for the diverse uses we can see so far, an EVP_SKEY would conceptually be a union of "a bunch of bytes" (with no particular algorithm in mind) and a reference to a provider-backed symmetric key (which will have to be specific to a key type / base cipher algo).
```

Reply:
BTW, this and the paragraph that follows belong in the "Libcrypto perspective" section

Reply:
Updated and reordered this and the following paragraphs, thanks!